k8s/cmd/nsmdp/nsmdp.go

```diff
@@ -231,6 +226,78 @@ func NewNSMDeviceServer(serviceRegistry serviceregistry.ServiceRegistry) error {
 	return err
 }
 
+func (nsm *nsmClientEndpoints) addClientDevice() {
+	nsm.mutext.Lock()
+	defer nsm.mutext.Unlock()
+
+	for {
+		nsm.clientId++
+
+		id := fmt.Sprintf("nsm-%d", nsm.clientId)
+		if _, ok := nsm.devs[id]; ok {
+			// Item already exists, increment to new one.
+			continue
+		}
+		// Add a new device Id
+		dev := &pluginapi.Device{
+			ID:     id,
+			Health: pluginapi.Healthy,
+		}
+		nsm.devs[id] = dev
+		nsm.resp.Devices = append(nsm.resp.Devices, dev)
+		break
+	}
+
+}
+
+func (n *nsmClientEndpoints) sendDeviceUpdate() {
+	n.mutext.Lock()
+	defer n.mutext.Unlock()
+	if n.pluginApi != nil {
+		api := *n.pluginApi
+		logrus.Infof("Send device plugins intfo %v", n.resp)
+		if err := api.Send(n.resp); err != nil {
+			logrus.Errorf("Failed to send response to kubelet: %v\n", err)
+		}
+	}
+}
+
+func (n *nsmClientEndpoints) reciveWorkspaces() {
+	for {
+		reply, err := enumWorkspaces(n.serviceRegistry)
+		if err != nil {
+			logrus.Errorf("Error receive devices from NSM. %v", err)
+			// Make a fast delay to faster startup of NSMD.
+			<-time.Tick(100 * time.Millisecond)
+			continue
```

Comment:
What are the chances we end up in an endless loop here?

Reply:
If NSMD will not respond properly, nothing will work in any case, so it is ok.